src/util/futurepool.rs

```diff
@@ -0,0 +1,258 @@
+// Copyright 2018 PingCAP, Inc.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+/// This mod implemented a wrapped future pool that supports `on_tick()` which is driven by
+/// tasks and is invoked no less than the specific interval.
+
+use std::fmt;
+use std::cell::{Cell, RefCell, RefMut};
+use std::sync::{mpsc, Arc};
+use std::thread;
+use std::time::Duration;
+use futures::Future;
+use futures_cpupool as cpupool;
+
+use util;
+use util::time::Instant;
+use util::collections::HashMap;
+
+pub trait Context: fmt::Debug + Send + 'static {
```

Comment:
Why a trait needs to implement `'static`?

Reply:
The context should lives long so that it is `'static`.

Reply:
I don't think it's a requirement for the trait. How long it needs to live depends on how you use the actual struct.

Reply:
Thanks, good point.

Reply:
I moved `'static` trait bound into `FuturePool` (i.e. `struct FuturePool<T: Context + 'static>`)